core/drivers/stm32_iwdg.c

```diff
@@ -0,0 +1,400 @@
+		panic();
+}
+
+static void iwdg_wdt_refresh(struct wdt_chip *chip)
+{
+	struct stm32_iwdg_device *iwdg = wdt_chip_to_iwdg(chip);
+
+	iwdg_refresh(iwdg);
+}
+
+static TEE_Result iwdg_wdt_set_timeout(struct wdt_chip *chip,
+				       unsigned long timeout)
+{
+	struct stm32_iwdg_device *iwdg = wdt_chip_to_iwdg(chip);
+
+	if (!iwdg_timeout_cnt(iwdg, timeout))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	iwdg->timeout = timeout;
+
+	if (is_enable(iwdg)) {
+		TEE_Result res = TEE_ERROR_GENERIC;
+
+		res = configure_timeout(iwdg);
+		if (res)
+			return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static const struct wdt_ops stm32_iwdg_ops = {
+	.start = iwdg_wdt_start,
+	.ping = iwdg_wdt_refresh,
+	.set_timeout = iwdg_wdt_set_timeout,
+};
+DECLARE_KEEP_PAGER(stm32_iwdg_ops);
+
+/* Refresh all registered watchdogs */
+void stm32_iwdg_refresh(void)
+{
+	struct stm32_iwdg_device *iwdg = NULL;
+	uint32_t exceptions = cpu_spin_lock_xsave(&iwdg_lock);
+
+	SLIST_FOREACH(iwdg, &iwdg_dev_list, link)
+		iwdg_refresh(iwdg);
+
+	cpu_spin_unlock_xrestore(&iwdg_lock, exceptions);
+}
+
+/* Driver initialization */
+static TEE_Result stm32_iwdg_parse_fdt(struct stm32_iwdg_device *iwdg,
+				       const void *fdt, int node)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct dt_node_info dt_info = { };
+	const fdt32_t *cuint = NULL;
+
+	_fdt_fill_device_info(fdt, &dt_info, node);
+
+	if (dt_info.reg == DT_INFO_INVALID_REG ||
+	    dt_info.reg_size == DT_INFO_INVALID_REG_SIZE)
+		panic();
+
+	res = clk_dt_get_by_name(fdt, node, "pclk", &iwdg->clock);
+	if (res)
+		return res;
+
+	res = clk_dt_get_by_name(fdt, node, "lsi", &iwdg->clk_lsi);
+	if (res)
+		return res;
+
+	if (dt_info.status == DT_STATUS_OK_NSEC)
+		iwdg->flags |= IWDG_FLAGS_NON_SECURE;
+
+	/* Get IOMEM address */
+	iwdg->base.pa = dt_info.reg;
+
+	if (iwdg->flags & IWDG_FLAGS_NON_SECURE)
+		io_pa_or_va_nsec(&iwdg->base, dt_info.reg_size);
+	else
+		io_pa_or_va_secure(&iwdg->base, dt_info.reg_size);
+
+	assert(iwdg->base.va);
+
+	/* Get and check timeout value */
+	cuint = fdt_getprop(fdt, node, "timeout-sec", NULL);
+	if (!cuint)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	iwdg->timeout = (int)fdt32_to_cpu(*cuint);
+	if (!iwdg->timeout)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!iwdg_timeout_cnt(iwdg, iwdg->timeout)) {
+		EMSG("Timeout %lu not applicable", iwdg->timeout);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	/* DT can specify low power cases */
+	if (!fdt_getprop(fdt, node, "stm32,enable-on-stop", NULL))
+		iwdg->flags |= IWDG_FLAGS_DISABLE_ON_STOP;
+
+	if (!fdt_getprop(fdt, node, "stm32,enable-on-standby", NULL))
+		iwdg->flags |= IWDG_FLAGS_DISABLE_ON_STANDBY;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_iwdg_setup(struct stm32_iwdg_device *iwdg,
+				   const void *fdt, int node)
+{
+	struct stm32_iwdg_otp_data otp_data = { };
+	TEE_Result res = TEE_SUCCESS;
+
+	res = stm32_iwdg_parse_fdt(iwdg, fdt, node);
+	if (res)
+		return res;
+
+	res = stm32_get_iwdg_otp_config(iwdg->base.pa, &otp_data);
```

Comment:
This function is added in a later commit.

Reply:
The driver expects the platform to provide this function, which is done by the next commit in the series.
A weak function in this source file is preferred?